src/object/schema_manager.c

```diff
@@ -15797,13 +15785,23 @@ sm_truncate_class_internal (MOP class_mop)
 	  continue;
 	}
 
-      if (keep_pk == true && c->type == SM_CONSTRAINT_PRIMARY_KEY)
+      if ((c->type == SM_CONSTRAINT_PRIMARY_KEY && classobj_is_pk_referred (class_mop, c->fk_info, false, NULL))
+	  || !sm_is_possible_to_recreate_constraint (class_mop, class_, c))
 	{
-	  /* Do not save PK referred by FK. the PK can't be dropped. */
-	  continue;
+	  /*
+	   * 1. the PK referred to by a FK.
+	   * 2. the B+tree which contains OIDs of other classes in the inheritance hierarchy.
+	   *
+	   * In these cases, We cannot drop and recreate the index as this might be
+	   * too costly, so we just remove the instances of the current class.
+	   */
+	  error = locator_remove_class_from_index (ws_oid (class_mop), &c->index_btid, &class_->header.ch_heap);
```

Comment:
How about moving this line out of for loop? I think it is more readable and It seems to serve the purpose of this for loop better.

Reply:
Do you mean like the original code?

Reply:
But, if done as you said, it'll still need a condition checking like this to skip a PK referred some FKs. Don't you think it looks redundant?

Reply:
I thought that a dummy variable which is the SM_CONSTRAINT_INFO type is used to collect constraints matched in these conditions. It looks a little in-efficient, but more readable. I think the comment just before for loop might be changed or removed if you don't want to change this processing(/* collect index information */).

Reply:
I see, I'll revise the comment.